<!DOCTYPE html>
<html >
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
      <title>Genome Collector</title>
    
      <link rel="stylesheet" href="_static/pygments.css">
      <link rel="stylesheet" href="_static/theme.css">
      <link rel="stylesheet" href="_static/sphinx_press_theme.css">
      
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>

      <!-- sphinx script_files -->
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>

      
      <script src="_static/theme-vendors.js"></script>
      <script src="_static/theme.js" defer></script>
    
  <link rel="index" title="Index" href="genindex.html" />
  <link rel="search" title="Search" href="search.html" />
  <link rel="next" title="Genome Collector API Reference" href="ref.html" /> 
  </head>

  <body>
    <div id="app" class="theme-container" :class="pageClasses"><navbar @toggle-sidebar="toggleSidebar">
  <router-link to="#" class="home-link">
    
      <span class="site-name">Genome Collector</span>
    
  </router-link>

  <div class="links">
    <navlinks class="can-hide">



    </navlinks>
  </div>
</navbar>

      
      <div class="sidebar-mask" @click="toggleSidebar(false)">
      </div>
        <sidebar @toggle-sidebar="toggleSidebar">
          
          <navlinks>
            



            
          </navlinks><div id="searchbox" class="searchbox" role="search">
  <div class="caption"><span class="caption-text">Quick search</span>
    <div class="searchformwrapper">
      <form class="search" action="search.html" method="get">
        <input type="text" name="q" />
        <input type="submit" value="Search" />
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div><div class="sidebar-links" role="navigation" aria-label="main navigation">
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="#more-biology-software">Reference</a></span>
      </p>
      <ul class="">
        
          <li class="toctree-l1 "><a href="ref.html" class="reference internal ">Genome Collector API Reference</a>

            
          </li>

        
      </ul>
    </div>
  
</div>
        </sidebar>

      <page>
          <div class="body-header" role="navigation" aria-label="navigation">
  
  <ul class="breadcrumbs">
    <li><a href="#">Docs</a> &raquo;</li>
    
    <li>Genome Collector</li>
  </ul>
  

  <ul class="page-nav">
  <li class="next">
    <a href="ref.html"
       title="next chapter">Genome Collector API Reference →</a>
  </li>
</ul>
  
</div>
<hr>
          <div class="content" role="main">
            
  <p align="center">
<img alt="Genome Collector Logo" title="DNA Chisel" src="https://raw.githubusercontent.com/Edinburgh-Genome-Foundry/genome_collector/master/docs/_static/images/title.png" width="600">
<br /><br />
</p><div class="section" id="genome-collector">
<h1>Genome Collector<a class="headerlink" href="#genome-collector" title="Permalink to this headline">¶</a></h1>
<a class="reference external image-reference" href="https://travis-ci.org/Edinburgh-Genome-Foundry/genome_collector"><img alt="Travis CI build status" src="https://travis-ci.org/Edinburgh-Genome-Foundry/genome_collector.svg?branch=master" /></a>
<a class="reference external image-reference" href="https://coveralls.io/github/Edinburgh-Genome-Foundry/genome_collector?branch=master"><img alt="https://coveralls.io/repos/github/Edinburgh-Genome-Foundry/genome_collector/badge.svg?branch=master" src="https://coveralls.io/repos/github/Edinburgh-Genome-Foundry/genome_collector/badge.svg?branch=master" /></a>
<p>Genome Collector (full documentation
<a class="reference external" href="https://edinburgh-genome-foundry.github.io/genome_collector/">here</a>)
is a Python library to download and manage reference genome data for specific
TaxIDs, in particular nucleotide and protein sequences (in fasta/genbank/gff
formats), and alignment databases (BLAST, Bowtie1/2).</p>
<p>The data is downloaded automatically on a need-to basis, making it very easy
for Python projects to use and re-use reference genomes of E. coli,
S. cerevisiae, and so on, without the worry of manually downloading from NCBI.</p>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<p>Let’s get Biopython records of all protein sequences in <em>E. coli</em>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">genome_collector</span> <span class="kn">import</span> <span class="n">GenomeCollection</span>
<span class="n">collection</span> <span class="o">=</span> <span class="n">GenomeCollection</span><span class="p">()</span>
<span class="n">records</span> <span class="o">=</span> <span class="n">collection</span><span class="o">.</span><span class="n">get_taxid_biopython_records</span><span class="p">(</span><span class="mi">511145</span><span class="p">,</span> <span class="s2">&quot;protein_fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>And that’s it! If the protein data wasn’t already on your machine, Genome
Collector downloaded from NCBI, and stored in your “collection” for the next
time time you need it.</p>
<p>Now let’s get a path to a local BLAST database for <em>S. cerevisiae</em>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">genome_collector</span> <span class="kn">import</span> <span class="n">GenomeCollection</span>
<span class="n">collection</span> <span class="o">=</span> <span class="n">GenomeCollection</span><span class="p">()</span>
<span class="n">db_path</span> <span class="o">=</span> <span class="n">collection</span><span class="o">.</span><span class="n">get_taxid_blastdb_path</span><span class="p">(</span><span class="n">taxid</span><span class="o">=</span><span class="mi">559292</span><span class="p">,</span> <span class="n">db_type</span><span class="o">=</span><span class="s1">&#39;nucl&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>If there was no <em>S. cerevisiae</em> database on your machine, Genome Collector
downloaded the genome data and built it. It is now in your collection, and you
can use the returned <code class="docutils literal notranslate"><span class="pre">db_path</span></code> to start a BLAST process:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">subprocess</span>
<span class="n">process</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span>
    <span class="s1">&#39;blastn&#39;</span><span class="p">,</span> <span class="s1">&#39;-db&#39;</span><span class="p">,</span> <span class="n">db_path</span><span class="p">,</span> <span class="s1">&#39;-query&#39;</span><span class="p">,</span> <span class="s1">&#39;queries.fa&#39;</span><span class="p">,</span> <span class="s1">&#39;-out&#39;</span><span class="p">,</span> <span class="s1">&#39;results.txt&#39;</span>
<span class="p">])</span>
</pre></div>
</div>
</div>
<div class="section" id="usage-tips">
<h2>Usage tips<a class="headerlink" href="#usage-tips" title="Permalink to this headline">¶</a></h2>
<div class="section" id="changing-the-data-storage-directory">
<h3>Changing the data storage directory<a class="headerlink" href="#changing-the-data-storage-directory" title="Permalink to this headline">¶</a></h3>
<p>You can decide where a collection’s local files will be stored with the
<code class="docutils literal notranslate"><span class="pre">data_dir</span></code> parameter of <code class="docutils literal notranslate"><span class="pre">GenomeCollection</span></code>. Note that the default value for
<code class="docutils literal notranslate"><span class="pre">data_dir</span></code> is highly recommended as it always points to the same local user
data folder. As a consequence, all librairies and applications using the
default will be able to pick genomes from the same folder. The path of this
default <code class="docutils literal notranslate"><span class="pre">collection.data_dir</span></code> is platform-specific:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">~/.local/share/genome_collector</span></code> on Linux</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">~/Library/Application</span> <span class="pre">Support/genome_collector</span></code> on MacOS</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">C:\Documents</span> <span class="pre">and</span> <span class="pre">Settings\&lt;User&gt;\Application</span> <span class="pre">Data\Local</span> <span class="pre">Settings\EGF\genome_collector</span></code> on Windows</p></li>
</ul>
<p>You can set the local default path globally at the beginning of your Python
script with:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">genome_collector</span> <span class="kn">import</span> <span class="n">GenomeCollection</span>
<span class="n">GenomeCollection</span><span class="o">.</span><span class="n">default_dir</span> <span class="o">=</span> <span class="s1">&#39;/my/new/dir&#39;</span>
</pre></div>
</div>
<p>Finally, you can set a default path as an environment variable (so it will be
shared by different Python processes):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">env</span> <span class="n">GENOME_COLLECTOR_DATA_DIR</span> <span class="o">=</span> <span class="o">/</span><span class="n">my</span><span class="o">/</span><span class="n">other</span><span class="o">/</span><span class="n">path</span>
</pre></div>
</div>
</div>
<div class="section" id="preventing-auto-download">
<h3>Preventing auto-download<a class="headerlink" href="#preventing-auto-download" title="Permalink to this headline">¶</a></h3>
<p>When using Genome Collector in a particular project, for instance a web app,
you may want to pre-download only a few genomes, and prevent users from using
other genomes. This can be done by setting a collection’s <code class="docutils literal notranslate"><span class="pre">autodownload</span></code>
attribute to False. To globally prevent Genome Collector from downloadind
data files, set this attribute at class level:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">GenomeCollection</span><span class="o">.</span><span class="n">autodownload</span> <span class="o">=</span> <span class="bp">False</span>
</pre></div>
</div>
</div>
<div class="section" id="command-line-interface">
<h3>Command line interface<a class="headerlink" href="#command-line-interface" title="Permalink to this headline">¶</a></h3>
<p>The very basic command-line interface enables to use Genome Collector to
pre-download genomes and pre-build BLAST databases on a machine. This can
be particularly useful in Dockerfiles to set up docker containers.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">genome_collector</span> <span class="n">genome</span> <span class="mi">511145</span>
<span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">genome_collector</span> <span class="n">blast_db</span> <span class="mi">511145</span> <span class="n">nucl</span>
</pre></div>
</div>
<p>By default these genomes will be downloaded to the platform-specific local
data folder. This can be changed by adding a data_dir at the end:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">genome_collector</span> <span class="n">genome</span> <span class="mi">511145</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">some</span><span class="o">/</span><span class="nb">dir</span><span class="o">/</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="similar-projects">
<h2>Similar projects<a class="headerlink" href="#similar-projects" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://github.com/kblin/ncbi-genome-download">ncbi-genome-download</a>
and <a class="reference external" href="https://github.com/simonvh/genomepy">Genomepy</a> are previous projects with
a similar goal. In comparison, Genome Collector is more opinionated, it uses
TaxID first, and is meant for use in Python scripts/Jupyter
notebooks (while the other tools are primarily for command-line usage), with
practical features like Biopython records loading and “on-demand”
genome downloading and database building.</p>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>You can install genome_collector through PIP</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sudo</span> <span class="n">pip</span> <span class="n">install</span> <span class="n">genome_collector</span>
</pre></div>
</div>
<p>Alternatively, you can unzip the sources in a folder and type</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sudo</span> <span class="n">python</span> <span class="n">setup</span><span class="o">.</span><span class="n">py</span> <span class="n">install</span>
</pre></div>
</div>
<p>For the BLAST-related features to work, you must have the NCBI BLAST software
installed. For instance on Ubuntu install with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sudo</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">install</span> <span class="n">ncbi</span><span class="o">-</span><span class="n">blast</span><span class="o">+</span>
</pre></div>
</div>
</div>
<div class="section" id="license-mit">
<h2>License = MIT<a class="headerlink" href="#license-mit" title="Permalink to this headline">¶</a></h2>
<p>genome_collector is an open-source software originally written at the
<a class="reference external" href="http://genomefoundry.org">Edinburgh Genome Foundry</a> by
<a class="reference external" href="https://github.com/Zulko">Zulko</a> and
<a class="reference external" href="https://github.com/Edinburgh-Genome-Foundry/genome_collector">released on Github</a>
under the MIT licence (copyright Edinburgh Genome Foundry).</p>
<p>Everyone is welcome to contribute !</p>
</div>
<div class="section" id="more-biology-software">
<h2>More biology software<a class="headerlink" href="#more-biology-software" title="Permalink to this headline">¶</a></h2>
<a class="reference external image-reference" href="https://edinburgh-genome-foundry.github.io/"><img alt="https://raw.githubusercontent.com/Edinburgh-Genome-Foundry/Edinburgh-Genome-Foundry.github.io/master/static/imgs/logos/egf-codon-horizontal.png" src="https://raw.githubusercontent.com/Edinburgh-Genome-Foundry/Edinburgh-Genome-Foundry.github.io/master/static/imgs/logos/egf-codon-horizontal.png" /></a>
<p>genome_collector is part of the <a class="reference external" href="https://edinburgh-genome-foundry.github.io/">EGF Codons</a> synthetic biology software suite for DNA design, manufacturing and validation.</p>
<div class="toctree-wrapper compound">
</div>
<div style='text-align: left; margin-bottom: 4em'>
    <a href="https://twitter.com/share" class="twitter-share-button"
    data-text="Genome Collector - Easy TaxID genome sequence and BLAST management with Python" data-size="large"
    data-hashtags="synbio,software">Tweet
    </a>
    <script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';
    if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+'://platform.twitter.com/widgets.js';
    fjs.parentNode.insertBefore(js,fjs);}}(document, 'script', 'twitter-wjs');
    </script>
    <iframe src="http://ghbtns.com/github-btn.html?user=Edinburgh-Genome-Foundry&repo=genome_collector&type=watch&count=true&size=large"
    allowtransparency="true" frameborder="0" scrolling="0" width="152px" height="30px" margin-bottom="30px"></iframe>
</div></div>
</div>


          </div>
          <div class="page-nav">
            <div class="inner"><ul class="page-nav">
  <li class="next">
    <a href="ref.html"
       title="next chapter">Genome Collector API Reference →</a>
  </li>
</ul><div class="footer" role="contentinfo">
      &#169; Copyright 2019, Edinburgh Genome Foundry.
    <br>
    Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0 with <a href="https://github.com/schettino72/sphinx_press_theme">Press Theme</a>.
</div>
            </div>
          </div>
      </page>
    </div>
    
    
  </body>
</html>