<!DOCTYPE html>
<html >
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
      <title>genome_collector.mixins.BlastMixin</title>
    
      <link rel="stylesheet" href="../../../_static/pygments.css">
      <link rel="stylesheet" href="../../../_static/theme.css">
      <link rel="stylesheet" href="../../../_static/sphinx_press_theme.css">
      
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>

      <!-- sphinx script_files -->
        <script type="text/javascript" src="../../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../../_static/language_data.js"></script>

      
      <script src="../../../_static/theme-vendors.js"></script>
      <script src="../../../_static/theme.js" defer></script>
    
  <link rel="index" title="Index" href="../../../genindex.html" />
  <link rel="search" title="Search" href="../../../search.html" /> 
  </head>

  <body>
    <div id="app" class="theme-container" :class="pageClasses"><navbar @toggle-sidebar="toggleSidebar">
  <router-link to="../../../index.html" class="home-link">
    
      <span class="site-name">Genome Collector</span>
    
  </router-link>

  <div class="links">
    <navlinks class="can-hide">



    </navlinks>
  </div>
</navbar>

      
      <div class="sidebar-mask" @click="toggleSidebar(false)">
      </div>
        <sidebar @toggle-sidebar="toggleSidebar">
          
          <navlinks>
            



            
          </navlinks><div id="searchbox" class="searchbox" role="search">
  <div class="caption"><span class="caption-text">Quick search</span>
    <div class="searchformwrapper">
      <form class="search" action="../../../search.html" method="get">
        <input type="text" name="q" />
        <input type="submit" value="Search" />
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div><div class="sidebar-links" role="navigation" aria-label="main navigation">
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="../../../index.html#more-biology-software">Reference</a></span>
      </p>
      <ul class="">
        
          <li class="toctree-l1 "><a href="../../../ref.html" class="reference internal ">Genome Collector API Reference</a>

            
          </li>

        
      </ul>
    </div>
  
</div>
        </sidebar>

      <page>
          <div class="body-header" role="navigation" aria-label="navigation">
  
  <ul class="breadcrumbs">
    <li><a href="../../../index.html">Docs</a> &raquo;</li>
    
      <li><a href="../../index.html">Module code</a> &raquo;</li>
    
    <li>genome_collector.mixins.BlastMixin</li>
  </ul>
  

  <ul class="page-nav">
</ul>
  
</div>
<hr>
          <div class="content" role="main">
            
  <h1>Source code for genome_collector.mixins.BlastMixin</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Mixin with BLAST methods, inherithed by GenomeCollection.&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">os</span>


<span class="k">class</span> <span class="nc">BlastMixin</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;All methods are directly accessible to GenomeCollection instances.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">generate_blast_db_for_taxid</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxid</span><span class="p">,</span> <span class="n">db_type</span><span class="o">=</span><span class="s2">&quot;nucl&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Generates a Blast DB for the TaxID. Autodownload FASTA if needed.</span>

<span class="sd">        ``db_type`` is either &quot;nucl&quot; (nucleotides database for blastn, blastx)</span>
<span class="sd">        or &quot;prot&quot; (protein database, untested).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">taxid</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">taxid</span><span class="p">)</span>
        <span class="n">data_type</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;nucl&quot;</span><span class="p">:</span> <span class="s2">&quot;genomic_fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;prot&quot;</span><span class="p">:</span> <span class="s2">&quot;protein_fasta&quot;</span><span class="p">}[</span><span class="n">db_type</span><span class="p">]</span>
        <span class="n">fa_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_taxid_genome_data_path</span><span class="p">(</span><span class="n">taxid</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">)</span>
        <span class="n">db_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datafile_path</span><span class="p">(</span><span class="n">taxid</span><span class="o">=</span><span class="n">taxid</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;blast_&quot;</span> <span class="o">+</span> <span class="n">db_type</span><span class="p">)</span>
        <span class="n">blast_args</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;makeblastdb&quot;</span><span class="p">,</span>
            <span class="s2">&quot;-in&quot;</span><span class="p">,</span>
            <span class="n">fa_path</span><span class="p">,</span>
            <span class="s2">&quot;-dbtype&quot;</span><span class="p">,</span>
            <span class="n">db_type</span><span class="p">,</span>
            <span class="s2">&quot;-out&quot;</span><span class="p">,</span>
            <span class="n">db_path</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">message</span> <span class="o">=</span> <span class="s2">&quot;Generating </span><span class="si">%s</span><span class="s2"> BLAST DB for taxid </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">db_type</span><span class="p">,</span> <span class="n">taxid</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_log_message</span><span class="p">(</span><span class="n">message</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">run_process</span><span class="p">(</span><span class="n">message</span><span class="p">,</span> <span class="n">blast_args</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_log_message</span><span class="p">(</span><span class="n">message</span> <span class="o">+</span> <span class="s2">&quot; - Done!&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_taxid_blastdb_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxid</span><span class="p">,</span> <span class="n">db_type</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get the path to a local blast DB, download and create one if needed.</span>

<span class="sd">        ``db_type`` is either &quot;nucl&quot; (nucleotides database for blastn, blastx)</span>
<span class="sd">        or &quot;prot&quot; (protein database, untested).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">taxid</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">taxid</span><span class="p">)</span>
        <span class="n">db_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datafile_path</span><span class="p">(</span><span class="n">taxid</span><span class="o">=</span><span class="n">taxid</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;blast_&quot;</span> <span class="o">+</span> <span class="n">db_type</span><span class="p">)</span>
        <span class="n">expected_file_extension</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;nucl&quot;</span><span class="p">:</span> <span class="s2">&quot;.nsq&quot;</span><span class="p">,</span> <span class="s2">&quot;prot&quot;</span><span class="p">:</span> <span class="s2">&quot;.psq&quot;</span><span class="p">}</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">db_path</span> <span class="o">+</span> <span class="n">expected_file_extension</span><span class="p">[</span><span class="n">db_type</span><span class="p">]):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">generate_blast_db_for_taxid</span><span class="p">(</span><span class="n">taxid</span><span class="p">,</span> <span class="n">db_type</span><span class="o">=</span><span class="n">db_type</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">db_path</span>

    <span class="k">def</span> <span class="nf">blast_against_taxid</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxid</span><span class="p">,</span> <span class="n">db_type</span><span class="p">,</span> <span class="n">blast_args</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Run a BLAST, using a genome_collector database.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ==========</span>

<span class="sd">        taxid</span>
<span class="sd">          TaxID (int or str) of the reference genome against which the query</span>
<span class="sd">          will be blaster. If no database for this taxID is available locally</span>
<span class="sd">          it will be downloaded.</span>

<span class="sd">        db_type</span>
<span class="sd">          Either &quot;nucl&quot; (nucleotides database for blastn, blastx) or &quot;prot&quot;</span>
<span class="sd">          (protein database, untested).</span>

<span class="sd">        blast_args</span>
<span class="sd">          List of NCBI-BLAST arguments, for instance [&#39;blastn&#39;, &#39;-query&#39;,</span>
<span class="sd">          &#39;my_sequences.fa&#39;, &#39;-out&#39;, &#39;myresults.xml&#39;].</span>

<span class="sd">        Examples</span>
<span class="sd">        ========</span>

<span class="sd">        &gt;&gt;&gt; blast_against_taxid(taxid, db_type, blast_args)</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">taxid</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">taxid</span><span class="p">)</span>
        <span class="n">db_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_taxid_blastdb_path</span><span class="p">(</span><span class="n">taxid</span><span class="o">=</span><span class="n">taxid</span><span class="p">,</span> <span class="n">db_type</span><span class="o">=</span><span class="n">db_type</span><span class="p">)</span>
        <span class="n">blast_args</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">blast_args</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;-db&quot;</span><span class="p">,</span> <span class="n">db_path</span><span class="p">]</span>
        <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;BLASTing against TaxID </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">: &quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">taxid</span><span class="p">,</span> <span class="n">db_type</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">run_process</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">blast_args</span><span class="p">)</span>
</pre></div>

          </div>
          <div class="page-nav">
            <div class="inner"><ul class="page-nav">
</ul><div class="footer" role="contentinfo">
      &#169; Copyright 2019, Edinburgh Genome Foundry.
    <br>
    Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0 with <a href="https://github.com/schettino72/sphinx_press_theme">Press Theme</a>.
</div>
            </div>
          </div>
      </page>
    </div>
    
    
  </body>
</html>