<!DOCTYPE html>
<html >
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
      <title>genome_collector.GenomeCollection</title>
    
      <link rel="stylesheet" href="../../_static/pygments.css">
      <link rel="stylesheet" href="../../_static/theme.css">
      <link rel="stylesheet" href="../../_static/sphinx_press_theme.css">
      
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>

      <!-- sphinx script_files -->
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>

      
      <script src="../../_static/theme-vendors.js"></script>
      <script src="../../_static/theme.js" defer></script>
    
  <link rel="index" title="Index" href="../../genindex.html" />
  <link rel="search" title="Search" href="../../search.html" /> 
  </head>

  <body>
    <div id="app" class="theme-container" :class="pageClasses"><navbar @toggle-sidebar="toggleSidebar">
  <router-link to="../../index.html" class="home-link">
    
      <span class="site-name">Genome Collector</span>
    
  </router-link>

  <div class="links">
    <navlinks class="can-hide">



    </navlinks>
  </div>
</navbar>

      
      <div class="sidebar-mask" @click="toggleSidebar(false)">
      </div>
        <sidebar @toggle-sidebar="toggleSidebar">
          
          <navlinks>
            



            
          </navlinks><div id="searchbox" class="searchbox" role="search">
  <div class="caption"><span class="caption-text">Quick search</span>
    <div class="searchformwrapper">
      <form class="search" action="../../search.html" method="get">
        <input type="text" name="q" />
        <input type="submit" value="Search" />
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div><div class="sidebar-links" role="navigation" aria-label="main navigation">
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="../../index.html#more-biology-software">Reference</a></span>
      </p>
      <ul class="">
        
          <li class="toctree-l1 "><a href="../../ref.html" class="reference internal ">Genome Collector API Reference</a>

            
          </li>

        
      </ul>
    </div>
  
</div>
        </sidebar>

      <page>
          <div class="body-header" role="navigation" aria-label="navigation">
  
  <ul class="breadcrumbs">
    <li><a href="../../index.html">Docs</a> &raquo;</li>
    
      <li><a href="../index.html">Module code</a> &raquo;</li>
    
    <li>genome_collector.GenomeCollection</li>
  </ul>
  

  <ul class="page-nav">
</ul>
  
</div>
<hr>
          <div class="content" role="main">
            
  <h1>Source code for genome_collector.GenomeCollection</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">subprocess</span>

<span class="kn">import</span> <span class="nn">proglog</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>

<span class="kn">from</span> <span class="nn">.mixins.BlastMixin</span> <span class="k">import</span> <span class="n">BlastMixin</span>
<span class="kn">from</span> <span class="nn">.mixins.NCBIMixin</span> <span class="k">import</span> <span class="n">NCBIMixin</span>
<span class="kn">from</span> <span class="nn">.mixins.FileManagerMixin</span> <span class="k">import</span> <span class="n">FileManagerMixin</span>
<span class="kn">from</span> <span class="nn">.mixins.BowtieMixin</span> <span class="k">import</span> <span class="n">BowtieMixin</span>


<div class="viewcode-block" id="GenomeCollection"><a class="viewcode-back" href="../../ref.html#genome_collector.GenomeCollection">[docs]</a><span class="k">class</span> <span class="nc">GenomeCollection</span><span class="p">(</span><span class="n">BlastMixin</span><span class="p">,</span> <span class="n">NCBIMixin</span><span class="p">,</span> <span class="n">FileManagerMixin</span><span class="p">,</span> <span class="n">BowtieMixin</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Collection of local data files including genomes and BLAST databases.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ==========</span>

<span class="sd">    data_dir</span>
<span class="sd">      Path to the directory where all genomes informations, sequences and</span>
<span class="sd">      blast database will be kept. The default value &quot;local_default&quot; means</span>
<span class="sd">      that the system-specific user data directory will be used, so either</span>
<span class="sd">      ``~/.local/share/genome_collector`` on Linux, or</span>
<span class="sd">      ``~/Library/Application Support/genome_collector`` on MacOS, etc.</span>

<span class="sd">    logger</span>
<span class="sd">      Logger to which the messages will be sent when downloading files or</span>
<span class="sd">      building blast databases. Use &quot;bar&quot; for a default bar logger, None</span>
<span class="sd">      for no logging, or any Proglog logger.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ==========</span>

<span class="sd">    autodownload</span>
<span class="sd">      When an absent local data file is requested and ``autodownload`` is True,</span>
<span class="sd">      the file will be automatically downloaded over the Internet. Otherwise</span>
<span class="sd">      a FileNotFoundError is raised. Attribute ``autodownload`` can be set to</span>
<span class="sd">      False on a single ``collection`` instance, or at class level with</span>
<span class="sd">      ``GenomeCollection.autodownload = False`` to globally prevent Genome</span>
<span class="sd">      Collector from downloadind data files.</span>

<span class="sd">    default_dir</span>
<span class="sd">      Default directory to use when parameter data_dir=&#39;default&#39;. This</span>
<span class="sd">      attribute allows to set a global default_dir at the beginning of a</span>
<span class="sd">      script with ``GlobalCollection.default_dir = &#39;/my/new/dir/&#39;``</span>

<span class="sd">    messages_prefix</span>
<span class="sd">      Prefix appearing as &quot;[prefix] &quot; in all logging messages.</span>

<span class="sd">    datafiles_extensions</span>
<span class="sd">      Dictionnary linking data file types to standardized file extensions.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">messages_prefix</span> <span class="o">=</span> <span class="s2">&quot;[genome_collector] &quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_dir</span><span class="o">=</span><span class="s2">&quot;default&quot;</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="s2">&quot;bar&quot;</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">data_dir</span> <span class="o">==</span> <span class="s2">&quot;default&quot;</span><span class="p">:</span>
            <span class="n">data_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_dir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data_dir</span> <span class="o">=</span> <span class="n">data_dir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_logger</span> <span class="o">=</span> <span class="n">proglog</span><span class="o">.</span><span class="n">default_bar_logger</span><span class="p">(</span><span class="n">logger</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_time_of_last_entrez_call</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">_log_message</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">message</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Send a message (with prefix) to the logger)&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_logger</span><span class="p">(</span><span class="n">message</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">messages_prefix</span> <span class="o">+</span> <span class="n">message</span><span class="p">)</span>

<div class="viewcode-block" id="GenomeCollection.get_taxid_infos"><a class="viewcode-back" href="../../ref.html#genome_collector.GenomeCollection.get_taxid_infos">[docs]</a>    <span class="k">def</span> <span class="nf">get_taxid_infos</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxid</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a dict with data about the taxid.</span>

<span class="sd">        Examples</span>
<span class="sd">        ========</span>

<span class="sd">        &gt;&gt;&gt; collection.get_taxid_infos(511145)</span>
<span class="sd">        &gt;&gt;&gt; {</span>
<span class="sd">        &gt;&gt;&gt;      &#39;Organism Name&#39;: &#39;Escherischia Coli&#39;, </span>
<span class="sd">        &gt;&gt;&gt;      &#39;DefLine&#39;: &#39;A well-studied enteric bacterium&#39;,</span>
<span class="sd">        &gt;&gt;&gt;      &#39;Organism_Kingdom&#39;: &#39;Bacteria&#39;,</span>
<span class="sd">        &gt;&gt;&gt;      &#39;AssemblyID&#39;: &#39;1755381&#39;,</span>
<span class="sd">        &gt;&gt;&gt;      ...</span>
<span class="sd">        &gt;&gt;&gt; }</span>



<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">taxid</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">taxid</span><span class="p">)</span>
        <span class="n">path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datafile_path</span><span class="p">(</span><span class="n">taxid</span><span class="o">=</span><span class="n">taxid</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;infos&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">autodownload</span><span class="p">:</span>
                <span class="n">error_message</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="s2">&quot;No infos for taxid </span><span class="si">%s</span><span class="s2"> found locally, and parameter &quot;</span>
                    <span class="s2">&quot;autodownload_enabled is set to False in &quot;</span>
                    <span class="s2">&quot;genome_collector.settings&quot;</span>
                <span class="p">)</span> <span class="o">%</span> <span class="n">taxid</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="n">error_message</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">download_taxid_genome_infos_from_ncbi</span><span class="p">(</span><span class="n">taxid</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span></div>


<div class="viewcode-block" id="GenomeCollection.get_taxid_genome_data_path"><a class="viewcode-back" href="../../ref.html#genome_collector.GenomeCollection.get_taxid_genome_data_path">[docs]</a>    <span class="k">def</span> <span class="nf">get_taxid_genome_data_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxid</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;genomic_fasta&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a path to the taxid&#39;s genome sequence. Download if needed.</span>
<span class="sd">        </span>
<span class="sd">        ``data_type`` is either genomic_fasta, genomic_genbank, genomic_gff,</span>
<span class="sd">        or protein_fasta</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">taxid</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">taxid</span><span class="p">)</span>
        <span class="n">path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datafile_path</span><span class="p">(</span><span class="n">taxid</span><span class="o">=</span><span class="n">taxid</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">autodownload</span><span class="p">:</span>
                <span class="n">error_message</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="s2">&quot;No genome for taxid </span><span class="si">%s</span><span class="s2"> found locally, and parameter &quot;</span>
                    <span class="s2">&quot;autodownload_enabled is set to False in &quot;</span>
                    <span class="s2">&quot;genome_collector.settings&quot;</span>
                <span class="p">)</span> <span class="o">%</span> <span class="n">taxid</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="n">error_message</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">download_taxid_genome_data_from_ncbi</span><span class="p">(</span>
                <span class="n">taxid</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">path</span></div>

<div class="viewcode-block" id="GenomeCollection.get_taxid_biopython_records"><a class="viewcode-back" href="../../ref.html#genome_collector.GenomeCollection.get_taxid_biopython_records">[docs]</a>    <span class="k">def</span> <span class="nf">get_taxid_biopython_records</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">taxid</span><span class="p">,</span> <span class="n">source_type</span><span class="o">=</span><span class="s2">&quot;genomic_genbank&quot;</span><span class="p">,</span> <span class="n">as_iterator</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a list of biopython records for the genome&#39;s chromosome.</span>
<span class="sd">        </span>
<span class="sd">        Even if there is a single record (for genomes with a single chromosome,</span>
<span class="sd">        a list is returned.</span>

<span class="sd">        For huge genomes, use the ``as_iterator`` option to return a Python</span>
<span class="sd">        iterator, which avoids to load all chromosomes at once in memory.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_taxid_genome_data_path</span><span class="p">(</span><span class="n">taxid</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="n">source_type</span><span class="p">)</span>
        <span class="n">data_format</span> <span class="o">=</span> <span class="n">source_type</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">records</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">data_format</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">as_iterator</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">records</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">records</span><span class="p">)</span></div>
    
    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">run_process</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">parameters</span><span class="p">):</span>
        <span class="n">process</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
            <span class="n">parameters</span><span class="p">,</span>
            <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
            <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">process</span><span class="o">.</span><span class="n">returncode</span><span class="p">:</span>
            <span class="n">error</span> <span class="o">=</span> <span class="n">process</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
            <span class="n">parameters</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">parameters</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">OSError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> failed:</span><span class="se">\n\n</span><span class="si">%s</span><span class="se">\n\n</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">error</span><span class="p">,</span> <span class="n">parameters</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">process</span><span class="o">.</span><span class="n">stdout</span></div>
</pre></div>

          </div>
          <div class="page-nav">
            <div class="inner"><ul class="page-nav">
</ul><div class="footer" role="contentinfo">
      &#169; Copyright 2019, Edinburgh Genome Foundry.
    <br>
    Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0 with <a href="https://github.com/schettino72/sphinx_press_theme">Press Theme</a>.
</div>
            </div>
          </div>
      </page>
    </div>
    
    
  </body>
</html>